src/Streamly/Internal/Data/Array/Foreign/Mut/Type.hs

```diff
@@ -1157,3 +1160,56 @@ toStreamD_ size Array{..} =
                     return r
         return $ D.Yield x (p `plusPtr` size)
 #endif
+
+{-# INLINE_NORMAL writeMaybesN #-}
+writeMaybesN :: forall m a. (MonadIO m, Storable a)
+    => Int -> Fold m (Maybe a) (Array a)
+writeMaybesN n = Fold step initial extract
+
+    where
+
+    initial = do
+        (Array start end _) <- liftIO $ newArray (max n 0)
+        return $ FL.Partial (ArrayUnsafe start end, 0 :: Int)
+
+    step (ArrayUnsafe start end, i :: Int) x = do
+        if isJust x && i < n -- check for boundary
+        then  do
+            liftIO $ poke end (fromJust x)
+            return $ FL.Partial (ArrayUnsafe start (end `plusPtr` sizeOf (undefined :: a)), i+1)
+        else do
+            return $ FL.Done $ Array start end end
+
+    extract (ArrayUnsafe start end , _) = return $ Array start end end -- liftIO . shrinkToFit
+
+{-# INLINE_NORMAL writeMaybes #-}
+writeMaybes :: forall m a. (MonadIO m, Storable a)
+    => Fold m (Maybe a) (Array a)
+writeMaybes = FL.rmapM extract $ FL.foldlM' step initial
+
+    where
+
+    alignSize = alignment (undefined :: a)
+
+    elemCount = bytesToElemCount (undefined :: a) (mkChunkSize 1024)
+
+    insertElem (Array start end bound) x = do
+        if isJust x
+        then  do
+            liftIO $ poke end (fromJust x)
+            return $ Array start (end `plusPtr` sizeOf (undefined :: a)) bound
+        else do
+            return $ Array start end bound
+
+
+    initial = do
+        when (elemCount < 0) $ error "toArrayMinChunk: elemCount is negative"
+        liftIO $ newArrayAligned alignSize elemCount
+    step arr@(Array start end bound) x | end == bound = do
+        let p = unsafeForeignPtrToPtr start
+            oldSize = end `minusPtr` p
+            newSize = max (oldSize * 2) 1
+        arr1 <- liftIO $ reallocAligned alignSize newSize arr
+        insertElem arr1 x
+    step arr x = insertElem arr x
+    extract = liftIO . shrinkToFit
```

Comment:
We should not be writing new custom combinators for this job. We can use existing fold/parse combinators.